src/Compiler/Microsoft.AspNetCore.Razor.Language/src/CodeGeneration/CodeWriter.cs

```diff
@@ -28,16 +42,62 @@ public CodeWriter(string newLine, RazorCodeGenerationOptions options)
         SetNewLine(newLine);
         IndentWithTabs = options.IndentWithTabs;
         TabSize = options.IndentSize;
-        _builder = new StringBuilder();
+
+        _indentSize = 0;
+        _indentString = ReadOnlyMemory<char>.Empty;
+
+        _pages = new();
+    }
+
+    private void AddChars(ReadOnlyMemory<char> value)
+    {
+        if (value.Length == 0)
+        {
+            return;
+        }
+
+        if (_pageIndex == _pages.Count)
+        {
+            _pages.Add(new ReadOnlyMemory<char>[PageSize]);
+        }
+
+        _pages[_pageIndex][_pageOffset] = value;
+        _pageOffset++;
+
+        if (_pageOffset == PageSize)
```

Comment:
Should this be `>=`? Is it not possible for `value` here to be longer than `PageSize - _pageOffset`, and there we'll be stuck on this page forever? Am I missing where the check for this is in `WriteCore`?

Reply:
No, it's not possible. `_pageOffset` never increments by more than one, and it isn't changed by any method but this one. The length of `value` doesn't have a relationship to `_pageOffset`. `_pageOffset` just represents the index of the array slot that `value` is copied into. We don't unpack the chars stored in `value`.

That said, it'd be good to add an assert that `_pageOffset` is always less than or equal to `PageSize`.

Reply:
ahh yes, my bad. Its a page or pages, not a page of content. I read PageSize wrong - its not the size of a page, its the size of a page or pages.

Reply:
No, not quite. A "page" is simply an array of `ReadOnlyMemory<char>`. Perhaps it's better to think of a "page" as an array of "chunks of text" and this method adds a new "chunk of text". "_pageIndex" is the index of the current page in the list of pages that we're maintaining. "_pageOffset" is the index into the current page.